### PR TITLE
Continuous benchmarking

### LINKED ISSUE #117: Continous Benchmarking
With https://github.com/rhysd/github-action-benchmark we can get https://rhysd.github.io/github-action-benchmark/dev/bench/ for our components.

Seems like a good investment to monitor for performance per-component.

### PR DESCRIPTION
**Pull request by echeran**
* Uses [Github Action for Continuous Benchmarking](https://github.com/rhysd/github-action-benchmark)
* Since we use Criterion.rs instead of the default Bencher crate in cargo for benchmarking, there were a couple of workarounds needed:
  * Use Criterion version >= 0.3.2 (we are using 0.3.3, the current latest)
  * Add a stanza in the component's Cargo.toml `[lib] bench = false` (see [issue Zibi opened](https://github.com/bheisler/criterion.rs/issues/433) to solicit help)
* Components lacking bench tests cause an error in the GH action's steps, so only components with bench tests are included
* Alert threshold can and should be tweaked
* `fail-on-alert: true` to make threshold alerts cause GH action failures on the PR
* Current way it works
  * Commits are made in the main repo in the `gh-pages` branch, but this is minimally impactful on the repo size, FWIW ([example](https://github.com/echeran/icu4x/commits/gh-pages))
  * Commits are copied over the docs repo, FWIW, also in the `gh-pages` branch ([example](https://github.com/echeran/icu4x-docs/commits/gh-pages))
  * The duplication can be resolved/consolidated in the future

Closes #117

### CODE REVIEW

File: .github/workflows/perf-runtime.yml [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      matrix:
+        component:
+          - locale
+          - num-util
+          - uniset
+      # Ensure that commits when pushing GH pages to external has linear commit
+      # history w/o merges (to keep things simpler). Do this via max-parallel=1.
+      # Also, necessary b/c can't figure out how to use component-specific branches
+      # and coordinate between the benchmark action step and the GH pages action step.
+      max-parallel: 1
```
Comment (zbraniecki) [inline .github/workflows/perf-runtime.yml:24]: do you think it's better over `cargo bench --all-features` from main `icu4x` directory?

File: .github/workflows/perf-runtime.yml [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+      - run: mkdir -p dev/${{ matrix.component }}/bench/cache
+
+      - name: Debug bench info
+        run: echo "ls -l dev" && ls -l dev && echo "ls -l dev/${{ matrix.component }}/bench" && ls -l dev/${{ matrix.component }}/bench
+      # - name: Debug cache info
+      #   run: echo "ls -l dev" && ls -l dev && echo "ls -l dev/${{ matrix.component }}/bench/cache" && ls -l dev/${{ matrix.component }}/bench/cache
+      #- run: mkdir -p dev/${{ matrix.component }}/bench/cache
+
+      - name: Try on a particular component
+        run: cd components/${{ matrix.component }}
```
Comment (zbraniecki) [inline .github/workflows/perf-runtime.yml:49]: what's the goal here?

**Comment by alerque** [comment]
I hate to rain on the parade and I'm seeing this PR pretty late, but I honestly don't think this approach is workable. I did some experimenting a while back trying to run benchmarks from CI (both in Actions and on Travis) and got wildly different results on different runs. A few runs in quick succession were often similar, but at different times of day or on different days they were not consistent enough to compare to each other. Most CI environments run in several layers of virtualization on shared hardware and are heavily subject to varying load conditions.

Additionally, coming back to the problem several weeks later I got not just subtle but wildly different numbers. Presumably my jobs got run in a different data center or on completely different hardware, and the difference can be substantial (more than 3× difference).

In other words logging the benchmark results and using them in the future is a recipe for disaster. In the absence of the hardware used and load at run time being predictable values in remote CI infrastructure, I suggest the only way to reasonably do benchmarks in CI is to run **both** the base used for comparison (whether that is the previous commit or the base commit of a branch) and the commit being compared sequentially in the same CI run in the same container. Then only evaluate the delta between these values, not the absolute value. The resulting absolute value will be meaningless to compare between runs, but the delta between two sequential runs in the same virtual machine and container environment will be reasonably meaningful.

File: .github/workflows/perf-runtime.yml (hunk 2 above)
Comment (echeran) [reply .github/workflows/perf-runtime.yml]: I think this is cruft from changing debugging steps.

**Comment by echeran** [comment]
@alerque Thanks, that's useful info.  I'm thinking that this PR could at least get automated dashboards set up with some data, and afterwards we can look at fixing things up like you're suggesting (connecting to a cloud provider that allows a controlled env and/or doing a prev+curr side-by-side comparison)

**Comment by alerque** [comment]
Also I haven't followed what the development workflow on this project looks like too closely but I hope the plan is to rewrite most of the commits and squash them into the few discrete changes relative to the branch base. :wink:

**Comment by echeran** [comment]
Yes, default workflow is adding commits (rebase discouraged -- it messes up comment history & reviews in Github), and squash merge.  I also prefer blanking out the commit comment beyond the summary first line since it seems redundant/noisy (the first line message contains the GH issue number already).

I think things got messy in the PR branch commit chain here because: 1) I don't have a way to test GH Action workflows locally which do stateful things and interact with external repos 2) the 5-10 minute latency on benchmark ops is a little distracting from workflow improvement work  3) not realizing the cache action has 2 distinct modes such that caching is necessary for the data file mode but caching interferes with the dashboard mode.

**Comment by echeran** [comment]
Ready to turn the PR from draft to active.

Changes since last comment:

- Commented out "comment-on-alert" configs for now b/c permissions on workflow runs on fork->upstream PRs not working
  * I didn't have permissions errors during my first level of personal testing, where the [testing PR](https://github.com/echeran/icu4x/pull/4)'s source and target branches are in the same repo (my personal fork)
  * This PR use the same source branch but the target here is upstream, so it's going cross-repo.  That's my best explanation for the [permission error](https://github.com/unicode-org/icu4x/runs/1030005223?check_suite_focus=true#step:4:16)
  * I referenced the [Benchmark GH Action's documentation](https://github.com/rhysd/github-action-benchmark) and comments on [when to create a personal access token](https://github.community/t5/GitHub-Actions/Github-action-not-triggering-gh-pages-upon-push/td-p/26869/highlight/false) and [how to do so](https://docs.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#permissions-for-the-github_token)
  * Should I be filing an upstream issue, or somewhere else?  It seems like some GH repo setting I'm not doing correctly
- Sorted out configs to get > 1 data point showing up (configs in and between benchmark action & copy GH pages action)
- Added links in the Readme to the dashboards that get created in the `icu4x-docs` repo, which is where our API docs are
  * Dashboards won't get created until the workflow runs, which only happens on merges to the `master` branch (same with API docs)
  * Based on tests in my personal fork ([example for num-util](https://echeran.github.io/icu4x-docs/dev/num-util/bench/))

**Comment by echeran** [comment]
cc @gnrunge FYI

**Review by zbraniecki — APPROVED** [review]

**Review by sffc — APPROVED** [review]
Looks good as a start for setting up the trendline dashboards.  I didn't review the code in great detail.
